### PR TITLE
CRT-TPC updates, additions

### PR DESCRIPTION
**Pull request by tboone0042**
Includes some updates made by Jacob Zettlemoyer and Gianluca Petrillo in one of Jacob's feature branches, focused on things needed for the trigger efficiency effort, namely the addition of the file `icaruscode/IcarusObj/CRTTPCMatchingInfo.h` and its inclusion in the related CRT-TPC code.

I've slightly expanded the matching algorithm scope to allow for better transparency into what it's doing so it can be evaluated. Includes the expansions to the "`matchCand`" structure laid out in the `icaruscode/CRT/CRTUtils/CRTT0MatchAlg.*` files as well as addition of an input boolean to the matching algorithm to allow for switching between compatibility with data and MC files (CRT Hit time is stored in different variables for data and MC).

This update also includes new piece of code called "`CRTTPCTruthEff_module.cc`" (along with its corresponding fcls) allows for generation of TTrees which can then be analyzed to understand what the matching algorithm is doing. 

More updates will come in future PRs as to the actual matching algorithm itself, however _this_ update is focused on getting the basic tools to evaluate it in place along with some of the general updates being made now that this code is being used for trigger efficiency.

### CODE REVIEW

**Comment by rennney** [comment]
trigger build

**Comment by tboone0042** [comment]
I'm not super familiar with the standard process here, do I need to figure out why this is failing at the build phase? It looks like there's a version conflict for art?

**Comment by rennney** [comment]
@tboone0042 Hi, yeh, I think it is unrelated to your work.

**Comment by tboone0042** [comment]
okay! I'll wait then, just making sure I'm not missing anything I'm supposed to be doing!

**Comment by rennney** [comment]
@tboone0042 Hi, I see a merge conflict, can you address it?

**Comment by tboone0042** [comment]
> @tboone0042 Hi, I see a merge conflict, can you address it?

On it! Give me a few minutes to double check that things still build if I remove those two lines

**Comment by tboone0042** [comment]
@rennney should hopefully be good now?

**Comment by tboone0042** [comment]
I got a few warnings in the build stage but nothing broke

**Comment by rennney** [comment]
trigger build

**Comment by rennney** [comment]
@tboone0042 Hi, please fix mostly unused variables errors from clang (c7). Please check if it builds locally with clang qualifiers

**Comment by tboone0042** [comment]
@rennney can you try again? I believe now I've fixed all the loose ends that it was complaining about...

**Comment by rennney** [comment]
trigger build

**Comment by rennney** [comment]
trigger build

**Comment by tboone0042** [comment]
do I need to fix any of the warnings?

File: icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc
Comment (PetrilloAtWork) [inline icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc:963]: Using `tracking::Point_t` as vector instead of `TVector3` _should_ avoid the need for some conversions.
Example later.

File: icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc
Comment (PetrilloAtWork) [inline icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc:987]: Defining `thisstart` as `Point_t` should allow this to be intuitively written as:
```suggestion
	geo::TPCID tpcid = fGeometryService->PositionToTPCID(thisstart);
	thisstart += fSCE->GetCalPosOffsets(thisstart,tpcid.TPC);
```

File: icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc
Comment (PetrilloAtWork) [inline icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc:1001]: As commented above, merging the common branch of these two `if` branches helps maintainability.

**Review by PetrilloAtWork — COMMENTED** [review]
I have left coding comments all across.
It's hard for me to spot issues in a code I don't know much of, but for what I see there is nothing clearly wrong.
I see that the interface has somewhere being extended to allow some data/MC specific behaviour. In some methods that hasn't happened and they are hard-coded to behave as data.
I would like a comment about the motivation and intention of that different treatment.

File: icaruscode/CRT/CRTTPCTruthEff_module.cc
Comment (tboone0042) [reply icaruscode/CRT/CRTTPCTruthEff_module.cc:93]: so this is the kind of thing I'm not sure what best practices are, I appreciate the note. do I have any easy way to rewrite this? I suppose putting what's in reconfigure into the beginning of the constructor, maybe?

File: icaruscode/CRT/CRTTPCTruthEff_module.cc
Comment (tboone0042) [reply icaruscode/CRT/CRTTPCTruthEff_module.cc:112]: I wasn't aware such a flag exists! I will try rewriting things with that in mind, that'd make things a _lot_ simpler

File: icaruscode/CRT/CRTTPCTruthEff_module.cc
Comment (tboone0042) [reply icaruscode/CRT/CRTTPCTruthEff_module.cc:555]: thanks, I copied this module from another one and didn't know if removing it would break things, I haven't had the time to experiment with that

File: icaruscode/CRT/CRTTPCTruthEff_module.cc
Comment (tboone0042) [reply icaruscode/CRT/CRTTPCTruthEff_module.cc]: I'm being told that the function "FindOneP" isn't a function in art?

File: icaruscode/CRT/CRTTPCTruthEff_module.cc
Comment (tboone0042) [reply icaruscode/CRT/CRTTPCTruthEff_module.cc:569]: This may be deprecated later, getCatCrossXYZ was just something I made to see what my own guess would compare to the one provided by the track but I don't really need it these days

File: icaruscode/CRT/CRTTPCTruthEff_module.cc
Comment (jzettle) [reply icaruscode/CRT/CRTTPCTruthEff_module.cc]: @tboone0042 Did you include the header in addition to making that change? There is a header for FindManyP `#include "canvas/Persistency/Common/FindManyP.h"` Try adding `#include "canvas/Persistency/Common/FindOneP.h"` and see if that helps

**Comment by tboone0042** [comment]
please do not merge while I'm working on these suggested edits, I've now broken everything and need to work through it

File: icaruscode/CRT/CRTTPCTruthEff_module.cc
Comment (tboone0042) [reply icaruscode/CRT/CRTTPCTruthEff_module.cc:94]: For now I've removed `reconfigure()` and put everything that was inside it into the constructor, hopefully that works!

File: icaruscode/CRT/CRTUtils/CRTT0MatchAlg.h
Comment (tboone0042) [reply icaruscode/CRT/CRTUtils/CRTT0MatchAlg.h:166]: I'm currently waiting to make sure these changes build, I'll be pushing this with the other notes soon

File: icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc
Comment (tboone0042) [reply icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc:461]: I don't use Emacs! I dunno why I removed them but I'm fine with them being in

Comment (jzettle) [reply icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc:461]: The marks do make Emacs unhappy but that may only affect a small number of people (I use Emacs at least). Not to say not to use them just a general comment

File: icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc
Comment (tboone0042) [reply icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc:859]: I got all of these inside the functions I use, there's a lot of deprecated code I should be excising that still uses the old method but for now I'm just changing what's useful

**Comment by tboone0042** [comment]
@PetrilloAtWork and @rennney can I get final approval of this? I got it to build normally but I keep getting weird errors not related to anything I've written when I try to build in Clang with v09_63_00, however I don't know what would've changed that would cause Clang issues since the last attempt

**Comment by rennney** [comment]
trigger build

**Comment by tboone0042** [comment]
I see the issues, fixing now

**Review by PetrilloAtWork — CHANGES_REQUESTED** [review]
There are a couple of bugs which rose from the changes (`fIsData` use in constructor and a `if(!t0)` test).
I have left a few more comments, but those are not essential.

File: icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc
Comment (tboone0042) [reply icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc]: so is passing by value better than by reference, say even if I made it `uint64_t const&`? Though now I'm beginning to recall that in one of your notes you did mention not to do the `const&` for "regular" data types, did I interpret that note correctly?

File: icaruscode/CRT/CRTTPCTruthEff_module.cc
Comment (tboone0042) [reply icaruscode/CRT/CRTTPCTruthEff_module.cc:351]: commented out for now

File: icaruscode/CRT/CRTTPCTruthEff_module.cc
Comment (tboone0042) [reply icaruscode/CRT/CRTTPCTruthEff_module.cc:400]: gotcha, i've made this change in what will be a commit soon, thanks!

File: icaruscode/CRT/CRTTPCTruthEff_module.cc
Comment (tboone0042) [reply icaruscode/CRT/CRTTPCTruthEff_module.cc:406]: thanks, will do the more explicit one, I appreciate the info!!

**Comment by tboone0042** [comment]
Hi @rennney could you try another build? I got everything to compile under both GCC and Clang on the GPVMs

**Comment by rennney** [comment]
trigger build

File: icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc
Comment (PetrilloAtWork) [reply icaruscode/CRT/CRTUtils/CRTT0MatchAlg.cc]: I confirm the recommendation
> not to do the `const&` for "regular" data types

**Comment by tboone0042** [comment]
okay @rennney I think it's ready for a final build check, I don't think any changes I've made will bring Gianluca's disapproval so if it builds here like it did for both "normal" and clang builds that I did just now then it should be ready for merging

**Comment by rennney** [comment]
trigger build

**Comment by PetrilloAtWork** [comment]
@tboone0042 you underestimate my pickiness!!

**Review by PetrilloAtWork — APPROVED** [review]
Despite myself, all looks good to me.

**Pull request merged**
